RecoParticleFlow/PFTracking/src/PFDisplacedVertexCandidateFinder.cc

```diff
@@ -198,24 +199,33 @@ void PFDisplacedVertexCandidateFinder::link(const TrackBaseRef& el1,
                                             double& dist,
                                             GlobalPoint& P,
                                             PFDisplacedVertexCandidate::VertexLinkTest& vertexLinkTest) {
```

Comment:
this is called from PFDisplacedVertexCandidateFinder::associate, which is itself called recursively; after that this method is also called in ::packLinks.
It's not obvious to me that the same el1,el2 link is tested only once.
If there are repeated calls there is some opportunity for optimization

Reply:
yes - I had that in mind with "algorithmic improvements". checking it now...

Reply:
I think there is a CPU-memory tradeoff here. `associate` calls `link` just once for each pair, but in order to run `packLinks`, you need to know how many elements were found in total. Therefore, one either needs to recompute the link at packing time (current approach) or cache `dist`, `GlobalPoint` and `linktest` for each pair until the end. Caching could be viable, but depends on the number of elements in a displaced candidate. I checked, and most have a few only, however candidates with thousands of tracks also exist (similar thing happens with linking if PFBlocks).

Therefore, I can't think of a way of improving this easily at the moment, without re-engineering the algorithm with potential physics changes. Unless I missed something simple...

Reply:
I'm guessing that O(10 MB) in-produce (something that goes away after ::produce exits) should be acceptable.  So, if it's thousands of track pairs then we are probably still OK, if it's thousands^2, it may be too much.

Reply:
The case is that `packLinks` is doing all pairs of tracks in the displaced candidate to build the distance matrix [link](https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFTracking/src/PFDisplacedVertexCandidateFinder.cc#L285) `associate` only considers a small number of tracks to build the candidate based on a linked list.

Therefore, caching `link` outputs won't be helpful here to save CPU time.

The bigger issue is that an event always seems to contain one pathological displaced candidate with a few thousand tracks. I'm not sure if this candidate results in anything useful down the line. Overall, I think some algorithmic improvements should be considered to avoid building these huge candidates, but this might require a closer look that physics performance is not reduced.

@bendavid @hatakeyamak what are your thoughts?